Improving playback performance and video sync

alexandrekreis opened the pull request:
- Popcorn "onFrame" method runs around 60 times per second. That was unnecessary , causing the CPU work to significantly increase, and preventing the two videos (deskshare and webcam) to run together smoothly

- We only synchronize the medias if one of them fires a "waiting" event. If the user pauses the playback, we synchronize them as well

- Pedro's refactoring

pedrobmarin committed:
Refactored shapes drawing

alexandrekreis committed:
Merge pull request #4 from pedrobmarin/shapes-drawing

Refactored shapes drawing

alexandrekreis committed:
Improving playback performance and video sync

- Popcorn "onFrame" method runs around 6 times per second. That was unnecessary , causing the CPU work to significantly increase, and preventing the two videos (deskshare and webcam) to run together smoothly

- We only synchronize the medias if one of them fires a "waiting" event. If the user pauses the playback, we synchronize them as well.

fcecagno commented on record-and-playback/presentation/playback/presentation/0.9.0/lib/writing.js:
```diff
@@ -386,64 +377,94 @@ function runPopcorn() {
-        if(!((p.paused() === true) && (p.seeking() === false))) {
+        if ( (!p.paused() || p.seeking()) && (p.currentTime() - lastFrameTime >= 0.1) ) {
+          lastFrameTime = p.currentTime();
           var t = p.currentTime().toFixed(1); //get the time and round to 1 decimal place
```
There are three calls to p.currentTime(), wouldn't be better to store it in a variable?
Also, we probably should do Math.abs(p.currentTime() - lastFrameTime) >= 0.1

fcecagno commented on record-and-playback/presentation/playback/presentation/0.9.0/playback.js:
```diff
@@ -379,13 +380,70 @@ load_deskshare_video = function () {
+       //here we sync medias
+       //setting currentTime on deskshare video will fire a 'canplaythrough' event after the seeking is done.
+       Popcorn("#deskshare-video").currentTime(Popcorn('#video').currentTime());
+   });
```
Is it possible to #deskshare-video be paused?

fcecagno commented on record-and-playback/presentation/playback/presentation/0.9.0/playback.js:
```diff
@@ -379,13 +380,70 @@ load_deskshare_video = function () {
+   Popcorn("#deskshare-video").on("canplayall", function() {
+      //when deskshare video is ready for the first time, sync times
+      Popcorn("#deskshare-video").currentTime(Popcorn('#video').currentTime());
+   });
```
What if #video isn't in "canplayall" mode?

fcecagno commented on record-and-playback/presentation/playback/presentation/0.9.0/playback.js:
```diff
@@ -379,13 +380,70 @@ load_deskshare_video = function () {
+            Popcorn('#video').play();
+         else console.log("deskshare-video canplaythrough | Not resuming playback: main media source not ready");
+      }
+   });
```
There's no canplaythrough for #deskshare-video, if it was in waiting state.

fcecagno commented on record-and-playback/presentation/playback/presentation/0.9.0/playback.js:
```diff
@@ -379,13 +380,70 @@ load_deskshare_video = function () {
+    if(!Popcorn('#video').paused() && !Popcorn('#video').seeking())
+        Popcorn("#video").pause();
+    else console.log("resync | main video seeking or already paused");
+}
```
Same thing here, we're only considering to enter resync after a waiting event on #deskshare-video.

fcecagno commented on record-and-playback/presentation/playback/presentation/0.9.0/playback.js:
```diff
@@ -379,13 +380,70 @@ load_deskshare_video = function () {
+     } else return 0;
+  } else return 0;
+}
+
```
Please refactor it, it doesn't look good.

var videoObject = $(videoTag)[0];
var videoBuffer = videoObject.buffered;

Why $(videoTag)[0]? Is it an array? What's in this array? How do you explain this while loop? Do we need to run it only for logging purposes?

alexandrekreis committed:
Video sync mechanism refactoring

- Created function "seySync( )" which sets the sync logic. Basically, we sync all medias when pausing the master video. We pause the master video when any video receives a waiting event, or when the master video is seeked. If necessary, we resume playing after the sync.

- Refactored when 'media-ready' event is fired.  We fire this event only when the medias receive a 'canplayall' event.

alexandrekreis commented:
Added sync mechanism refactoring.
